Create package cloudinfo

AnanyaKumar opened the pull request:
Started implementing the cloudinfo package. This is my first time writing production Go code, so I'm sending a short (50 line) diff for feedback. Thanks!

AnanyaKumar commented:
I authored these commits.

vishh commented:
ok to test

On Wed, Jun 10, 2015 at 3:09 PM, cadvisorJenkinsBot <
notifications@github.com> wrote:

> Can one of the admins verify this patch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/cadvisor/pull/769#issuecomment-110931799>.
>

vishh commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,56 @@
+	GCE CloudProvider = iota
+	AWS
+	BAREMETAL
+	OTHER
```
Could this be "unknown" instead?

AnanyaKumar replied:
Done.

vishh commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+func inGCE() bool {
+	_, err := http.Get("http://metadata.google.internal/" +
```
nit: The line wrap isn't necessary.

AnanyaKumar replied:
Removed.

vishh commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,56 @@
+	OTHER
+)
+
+func GetCloudProvider() CloudProvider {
```
Consider creating an interface for this method. It will be useful for unit testing.

AnanyaKumar replied:
Added.

vishh commented:
@AnanyaKumar: This PR seems to be mostly fine.

AnanyaKumar commented:
Thanks for your feedback @vishh! I've made the changes, and I'll send a PR with more changes later on :)

rjnagal commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,96 @@
+	return &realCloudInfo{}, nil
+}
+
+func (self *realCloudInfo) GetCloudProvider() (CloudProvider, error) {
```
I'd suggest we split this method into two:

a private detect() method that detects which cloudprovider the node is running on. This can be called once in NewRealCloudInfo() and we can save the result as this isn't going to change across the machine's lifetime.

The public part of the interface should just return the detected type rather than calling detection logic on every invocation. We can also cache instance type as that's mostly the same detect call.

AnanyaKumar replied:
Great idea! Changed the code.

rjnagal commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,96 @@
+	return UNKNOWN, nil
+}
+
+func (self *realCloudInfo) GetInstanceType(cp CloudProvider) (string, error) {
```
If we save cloudprovider and instance type internally, we don't need to accept cloudprovider as a parameter.

AnanyaKumar replied:
True!

rjnagal commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,96 @@
+	return NOINSTANCE, nil
+}
+
+func inGCE() bool {
```
Its likely we'd would expand cloudprovider utils to return more than just instance type. I'd suggest moving cloudprovider logic to its own file for supported clouds (gce for now). The default/unimplemented ones can stay here.

AnanyaKumar replied:
That's true - changed!

AnanyaKumar commented:
The feature has been implemented. Couple of notes:
1. I have tested this on my local machine and it works as expected. I'm in the process of testing it in GCE (I'm installing cAdvisor on a GCE instance right now)
2. Errors were, for the most part, removed. If there was an error in getting the cloud provider, then the cloud provider will be set to "UNKNOWN", so the errors are contained in the return value.
3. I've changed the type of the constants to strings, so that they're more readable in the API output.

Let me know if you have any feedback!

rjnagal commented on info/v1/machine.go:
```diff
@@ -14,6 +14,10 @@
 
+import (
+	"github.com/google/cadvisor/utils/cloudinfo"
+)
```
I'd suggest moving the datastructures reported in the API  (CloudProvider, InstanceType) here. That way, consumers of the api don't have to include utils/cloudinfo.

AnanyaKumar replied:
Sorry, I didn't understand why the data structures should go into machine.go Why shouldn't consumers of the API have to include utils/cloudinfo? Aren't the methods that retrieve information about the cloud provider in utils/cloudinfo anyway? Thanks!

rjnagal commented:
minor code movement.
LGTM, otherwise.

AnanyaKumar commented:
I didn't detect the instance type correctly in GCE in the previous commit. I've corrected this, and tested on an n1-standard-1 instance (it outputs the correct instance type).

AnanyaKumar commented:
@rjnagal Does this look fine?

rjnagal commented:
I think this is fine.
Can you squash the commits?

AnanyaKumar commented:
@rjnagal Awesome, thanks for your feedback. I've squashed the commits!

vmarmol commented on info/v1/machine.go:
```diff
@@ -112,6 +112,22 @@ type NetInfo struct {
+const (
+	GCE              CloudProvider = "GCE"
+	AWS                            = "AWS"
+	BAREMETAL                      = "BAREMETAL"
```
nit: Go style is camelcase, this is fine for the above, but for the others it should be Baremetal and UnknownProvider

vmarmol commented on info/v1/machine.go:
```diff
@@ -112,6 +112,22 @@ type NetInfo struct {
+type InstanceType string
+
+const (
+	NO_INSTANCE      InstanceType = "NONE"
```
nit: Same here

vmarmol commented on info/v1/machine.go:
```diff
@@ -143,6 +159,12 @@ type MachineInfo struct {
+	CloudProvider CloudProvider `json:"cloud_provider"`
+
+	// Type of cloud instance (e.g. GCE standard) the machine is.
+	InstanceType InstanceType `json:"instance_type"`
```
nit: omitempty that way it is not included when it is not used. Same for CloudProvider:

```
`json:"instance_type,omitempty"`
```

vmarmol replied:
Although I guess we have Unknown which will make this never happen :P ignore...

vmarmol commented on utils/cloudinfo/cloudinfo.go:
```diff
@@ -0,0 +1,87 @@
+	instanceType  info.InstanceType
+}
+
+func NewRealCloudInfo() CloudInfo {
```
nit: Just `NewCloudInfo()`

vmarmol commented on utils/cloudinfo/gce.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+func inGCE() bool {
+	_, err := http.Get("http://metadata.google.internal/computeMetadata/v1/instance/machine-type")
```
I recommend using: https://godoc.org/google.golang.org/cloud/compute/metadata

It's pretty simple code, but useful to use a library by the authors/maintainers :)

AnanyaKumar commented:
@vmarmol Sorry I missed your comments until you spoke to me IRL a couple of days ago. I've pushed a diff based on the corrections to the code that you suggested. I'll respond faster if you have any follow ups!

vmarmol commented:
LGTM, thanks @AnanyaKumar!
